### ISSUE TITLE
Why doesn't line352 of attention_decoder.py need to reuse the variable of this function like line347?

### ISSUE THREAD

**Issue by alanlisten** [yaserkl/RLSeq2Seq#18]
if i == 0 and initial_state_attention:  # always true in decode mode
        with variable_scope.variable_scope(variable_scope.get_variable_scope(), reuse=True): 
        ...
else:
       # should here add "with variable_scope ...., reuse=True): "?
        context_vector, attn_dist, coverage, masked_e = attention(state, tf.stack(temporal_e,axis=0), coverage)

**Comment by yaserkl** [comment]
I'd suggest to remove i==0 and see what error you'll get.
Hint: The error that you'll see will happen since in the first iteration (i=0) the required trainable variables will be defined, however in following steps, you should only use these trainable variable not defining them again (if you don't have this if, model will try to define these variables again at step i=1 and that's when you receive an error saying the "XXX" variable has been already defined and exist in the graph). That's why you need the reuse=True parameter to simply reuse already defined variable after the first step.

**Comment by alanlisten** [comment]
Ok, I understand what you mean. It is because the attention function has already been run on line 321 before the start of the loop, so in order to continue using the variable of the attention function, it is necessary to indicate the reuse in i=0. However, since the variable created in the attention function is encapsulated into the function, it is impossible to make reuse=Ture in the scope（named “Attention”） of the function. So you can only explicitly state that all variables under the scope of the previous level (named “attention_decoder”) are reused.
In other words, when i=0, the variables that indicate the attention function can be reused in the graph. In the following loop, Tensorflow will continue to reuse the variables of the attention function, and there is no need to write reuse=True again（But I think it’s okay to write again）.
Am I correct?

**Comment by yaserkl** [comment]
No, don't think of it like encapsulation in python. The tensorflow graph is accessible throughout the whole program and you can make changes in everywhere throughout your program. So, once you define a specific variable in the graph even inside a function, you can still access it from outside that function. Just try to write a simple sample and you'll understand it.

**Comment by alanlisten** [comment]
I understand, thank you for your patience. O(∩_∩)O
